test/e2e/runner/stats.go

```diff
@@ -0,0 +1,233 @@
+	jsn, err := json.Marshal(t)
+	if err != nil {
+		fmt.Errorf("failed to marshal state: %w")
+		return ""
+	}
+	return string(jsn)
+}
+
+func (t *Stats) String() string {
+	return fmt.Sprintf(`bandwidth="%v", added="%v"`, t.bandwidth, t.added)
+}
+
+func (t *Stats) TotalBandwidth(testnet *e2e.Testnet) int {
+	total := 0
+	for _, n := range testnet.Nodes {
+		for _, m := range testnet.Nodes {
+			total += t.bandwidth[n][m]
+		}
+	}
+	return total
+}
+
+func (t *Stats) TxsAdded(testnet *e2e.Testnet) float32 {
+	count := 0
+	for _, n := range testnet.Nodes {
+		count += t.added[n]
+	}
+	return float32(count) / float32(len(testnet.Nodes))
+}
+
+func (t *Stats) txsSent(testnet *e2e.Testnet) float32 {
+	rtotal := float32(0)
+	for _, n := range testnet.Nodes {
+		rtotal += float32(t.sent[n])
+	}
+	return rtotal / float32(len(testnet.Nodes))
+}
+
+func (t *Stats) Completion(testnet *e2e.Testnet, txsSent int) float32 {
+	total := float32(0)
+	for _, n := range testnet.Nodes {
+		total += float32(t.added[n]) / float32(txsSent)
+	}
+	return float32(100) * (total / float32(len(testnet.Nodes)))
+}
+
+func (t *Stats) BandwidthGraph(testnet *e2e.Testnet, computeRatio bool) map[string]map[string]float32 {
+	result := map[string]map[string]float32{}
+	for i, n := range testnet.Nodes {
+		sum := 1
+		if computeRatio {
+			sum := 0
+			for _, m := range testnet.Nodes {
+				sum += t.bandwidth[n][m]
+			}
+		}
+		result[strconv.Itoa(i)] = map[string]float32{}
+		for j, m := range testnet.Nodes {
+			result[strconv.Itoa(i)][strconv.Itoa(j)] = float32(t.bandwidth[n][m]) / float32(sum)
+		}
+	}
+	return result
+}
+
+func (t *Stats) Duplicates(testnet *e2e.Testnet) map[string]int {
+	result := map[string]int{}
+	for i, n := range testnet.Nodes {
+		result[strconv.Itoa(i)] = t.redundant[n]
+	}
+	return result
+}
+
+func (t *Stats) Peers(testnet *e2e.Testnet) map[string]int {
+	result := map[string]int{}
+	for i, n := range testnet.Nodes {
+		result[strconv.Itoa(i)] = t.peers[n]
+	}
+	return result
+}
+
+func (t *Stats) Redundancy(testnet *e2e.Testnet) float32 {
+	rtotal := float32(0)
+	stotal := float32(0)
+	for _, n := range testnet.Nodes {
+		rtotal += float32(t.redundant[n])
+		stotal += float32(t.added[n])
+	}
+	return rtotal / stotal
+}
+
+func (t *Stats) Degree(testnet *e2e.Testnet) float32 {
+	rtotal := float32(0)
+	for _, n := range testnet.Nodes {
+		rtotal += float32(t.peers[n])
+	}
+	return rtotal / float32(len(testnet.Nodes))
+}
+
+func (t *Stats) CPULoad(testnet *e2e.Testnet) float32 {
+	count := float32(0)
+	for _, n := range testnet.Nodes {
+		count += t.cpuLoad[n]
+	}
+	return count / float32(len(testnet.Nodes))
+}
+
+func queryInt(v1api v1.API, timeout time.Duration, field string, node string, extra string) (int, error) {
+	if result, err := doQuery(v1api, timeout, field, node, extra); err == nil {
+		if len(result.(model.Vector)) != 0 {
+			return strconv.Atoi(result.(model.Vector)[0].Value.String())
+		} else {
+			return 0, nil
+		}
```

Comment:
Ugly, but the go way.

```suggestion
		}
		return 0, nil
```

Reply:
Fixed. In fact, I prefer this pattern :)